File: srcp/2021/day3.py
from srcp.utils.input import get_data

data: list[str] = get_data(day=3, year=2021).splitlines()


def part_1(data: list[str]) -> int:
    length: int = len(data[0])
    gamma = []
    epsilon = []

    for l in range(0, length):
        one_count = 0
        zero_count = 0

        for line in data:
            tmp = line[l]
            if tmp == '1':
                one_count += 1
            else:
                zero_count += 1
        if one_count > zero_count:
            gamma.append(1)
            epsilon.append(0)
        else:
            gamma.append(0)
            epsilon.append(1)

    result_gamma = "".join([str(x) for x in gamma])
    result_epsilon = "".join([str(x) for x in epsilon])
    return int(result_gamma, 2) * int(result_epsilon, 2)


print(f"Partie 1 : {part_1(data)}")

File: srcr/src/days_2021/day3.rs

pub fn solve_day3(input: &str) -> u32 {
    let lines: Vec<&str> = input.lines().collect();

    let length: usize = lines[0].len();

    let mut gamma: Vec<usize> = vec![0; length];
    let mut epsilon: Vec<usize> = vec![0; length];

    for index in 0..length {
        let mut one_count: usize = 0;
        let mut zero_count: usize = 0;

        for line in lines.iter() {
            let tmp_val = line.chars().nth(index).unwrap();
            if tmp_val == '1' {
                one_count += 1;
            }
            if tmp_val == '0' {
                zero_count += 1;
            }
        }

        if one_count > zero_count {
            gamma[index] = 1;
            epsilon[index] = 0;
        }else{
            gamma[index] = 0;
            epsilon[index] = 1;
        }
    }

    let result_gamma = gamma.iter().map(|&x| x.to_string()).collect::<Vec<String>>().join("");

    let mut total_part_1 = 0;


    match i32::from_str_radix(&result_gamma, 2) {
        Ok(x) => {
            println!("result gamma: {}", x);
            total_part_1 = x;
        },
        Err(e) => println!("Erreur gamma: {}", e),
    }

    let result_epsilon = epsilon.iter().map(|&x| x.to_string()).collect::<Vec<String>>().join("");

    match i32::from_str_radix(&result_epsilon, 2) {
        Ok(x) => {
            println!("result epsilon: {}", x);
            total_part_1 = total_part_1 * x;
        },
        Err(e) => println!("Erreur epsilon: {}", e),
    }

    println!("result part 1: {}", total_part_1);

    0
}